internal/app/splinter/api/api.go

```diff
@@ -0,0 +1,263 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/gorilla/mux"
+	"github.com/mp-hl-2021/splinter/internal/app/splinter/model"
+	"log"
+	"net/http"
+	"reflect"
+	"strings"
+)
+
+type Api struct {
+	model.UserInterface
+}
+
+func NewApi(x model.UserInterface) Api {
+	return Api{x}
+}
+
+type Context struct {
+	*Api
+	wr  http.ResponseWriter
+	req *http.Request
+}
+
+type ErrorResponse struct {
+	Error      string
+	StatusCode int
+}
+
+func (ctx *Context) abortf(statusCode int, format string, a ...interface{}) {
+	ctx.wr.WriteHeader(statusCode)
+	t := ErrorResponse{
+		fmt.Errorf(format, a...).Error(),
+		statusCode,
+	}
+
+	if err := json.NewEncoder(ctx.wr).Encode(t); err != nil {
+		log.Fatal(err)
+	}
+
+	panic(t)
+}
+
+func (ctx *Context) abort(err error) {
+	ctx.abortf(http.StatusBadRequest, "%s", err.Error())
+}
+
+func (ctx *Context) token() model.Token {
+	header := ctx.req.Header.Get("Authorization")
+	if strings.HasPrefix(header, "Bearer ") {
+		return model.Token(strings.TrimPrefix(header, "Bearer "))
+	} else {
+		return model.UnauthenticatedToken
+	}
+}
+
+func (ctx *Context) decodeJson(v interface{}) {
+	if ctx.req.Header.Get("Content-Type") != "application/json" {
+		ctx.abortf(http.StatusBadRequest, "expected json data")
+	}
+
+	if err := json.NewDecoder(ctx.req.Body).Decode(v); err != nil {
+		ctx.abortf(http.StatusBadRequest, "invalid json data: %e", err)
+	}
+}
+
+func (ctx *Context) writeJson(v interface{}) {
+	ctx.wr.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(ctx.wr).Encode(v); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (ctx *Context) ok() {
+	ctx.wr.WriteHeader(http.StatusNoContent)
+}
+
+func (ctx *Context) do(err error) {
+	if err != nil {
+		ctx.abort(err)
+	}
+	ctx.ok()
+}
+
+func (a *Api) MakeHandler(name string) http.Handler {
```

Comment:
It is an interesting and cool way to do handlers. But could you explain the WHY of this solution?
For me it just contradicts the philosophy of Go. If you are trying to reduce the number of lines, you could use middleware technique and place it around entire router.